subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java

```diff
@@ -38,15 +46,28 @@
     private final Convention _convention;
     private final IConventionAware _source;
     private final Set<String> _propertyNames;
+    private final Map<String, Boolean> _propertyNamesAsProviderMapping;
     private final Map<String, MappedPropertyImpl> _mappings = new HashMap<String, MappedPropertyImpl>();
 
     public ConventionAwareHelper(IConventionAware source, Convention convention) {
         this._source = source;
         this._convention = convention;
+        this._propertyNamesAsProviderMapping = ClassInspector.inspect(source.getClass()).getProperties().stream().collect(Collectors.toMap(it -> it.getName(), it -> isProviderType(it.getGetters())));
         this._propertyNames = JavaPropertyReflectionUtil.propertyNames(source);
     }
 
+    private static boolean isProviderType(List<Method> getters) {
+        return getters.stream().anyMatch(ConventionAwareHelper::isProviderType);
+    }
+
+    private static boolean isProviderType(Method getter) {
+        return Property.class.isAssignableFrom(getter.getReturnType()) || ListProperty.class.isAssignableFrom(getter.getReturnType()) || SetProperty.class.isAssignableFrom(getter.getReturnType()) || MapProperty.class.isAssignableFrom(getter.getReturnType());
```

Comment:
I think this should be just `Provider.class.isAssignableFrom(getter.returnType)`. We don't want convention mappings to be applied to any kind of `Provider`.

Reply:
The convention mapping works for `Provider` but is completely ignored on `Property` types. This PR is strictly addressing the `Property` and throws an exception. For the other type, I was going to issue a deprecation warning as it works properly but we want to stop people from using them as a separate PR.